de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/allocation/transfer/HUTransformService.java

```diff
@@ -986,4 +989,133 @@ public IMutableAllocationResult process(final IHUContext localHuContext)
 		Preconditions.checkState(!productStorages.isEmpty(), "The list of productStorages below the given 'tuHU' may not be empty; tuHU=%s", tuHU);
 		return productStorages;
 	}
+
+	@lombok.Value
+	public static class HUsToNewCUsRequest
+	{
+		List<I_M_HU> sourceHUs;
+
+		Quantity qtyCU;
+
+		public static HUsToNewCUsRequest forSourceHuAndQty(@NonNull I_M_HU sourceHU, @NonNull Quantity qtyCU)
+		{
+			return HUsToNewCUsRequest.builder().sourceHU(sourceHU).qtyCU(qtyCU).build();
+		}
+
+		@lombok.Builder
+		private HUsToNewCUsRequest(
+				@Singular("sourceHU") @NonNull final List<I_M_HU> sourceHUs,
+				@NonNull final Quantity qtyCU)
+		{
+			this.sourceHUs = sourceHUs;
+
+			this.qtyCU = qtyCU;
+		}
+	}
+
+	public List<I_M_HU> husToNewCUs(@NonNull final HUsToNewCUsRequest newCUsRequest)
+	{
+		Quantity qtyCuLeft = newCUsRequest.getQtyCU();
+
+		final ImmutableList.Builder<I_M_HU> result = ImmutableList.builder();
+		for (final I_M_HU sourceHU : newCUsRequest.getSourceHUs())
+		{
+			final List<I_M_HU> currentResult = huToNewCUs(sourceHU, qtyCuLeft);
+			result.addAll(currentResult);
+
+			qtyCuLeft = qtyCuLeft.subtract(calculateTotalQtyOfCUs(currentResult, qtyCuLeft.getUOM()));
+
+			if (qtyCuLeft.signum() <= 0)
+			{
+				break;
+			}
+		}
+
+		return result.build();
+	}
+
+	private List<I_M_HU> huToNewCUs(@NonNull final I_M_HU sourceHU, final Quantity qtyCU)
+	{
+		if (handlingUnitsBL.isLoadingUnit(sourceHU))
+		{
+			return luExtractCUs(sourceHU, qtyCU);
+		}
+		else if (handlingUnitsBL.isTransportUnitOrAggregate(sourceHU))
+		{
+			return tuExtractCUs(sourceHU, qtyCU);
+
+		}
+		else
+		{
+			return cuToNewCU(sourceHU, qtyCU);
+		}
+	}
+
+	private List<I_M_HU> luExtractCUs(@NonNull final I_M_HU sourceLU, final Quantity qtyCU)
+	{
+		Preconditions.checkArgument(qtyCU.signum() > 0, "qtyCU > 0");
+
+		final ImmutableList.Builder<I_M_HU> extractedCUs = new ImmutableList.Builder<>();
+
+		Quantity qtyCUsRemaining = qtyCU; // how many CUs we still have to extract
+
+		for (int i = 0; i < handlingUnitsDAO.retrieveIncludedHUs(sourceLU).size(); i++)
```

Comment:
the `handlingUnitsDAO.retrieveIncludedHUs(sourceLU).size()` is eavluated on each iteration of the for loop
* performance
* i think due to the `luExtractTUs` that we have within the loop, it will be decreased by one each time.

Reply:
You're right. Fixed by extracting in in a variable beforehand.